src/main/java/com/couchbase/client/dcp/conductor/Conductor.java

```diff
@@ -85,8 +89,27 @@ public SessionState sessionState() {
     }
 
     public Completable connect() {
-        Completable atLeastOneConfig = configProvider.configs().first().toCompletable();
-        return configProvider.start().concatWith(atLeastOneConfig);
+        return Observable.just(configProvider)
+                .flatMap(new Func1<ConfigProvider, Observable<?>>() {
+                             @Override
+                             public Observable<?> call(ConfigProvider configProvider) {
+                                 Completable atLeastOneConfig = configProvider.configs().first()
+                                         .timeout(env.bootstrapTimeout(), TimeUnit.SECONDS).toCompletable();
+                                 return configProvider.start().concatWith(atLeastOneConfig).toObservable();
+                             }
+                         }
+                ).retryWhen(anyOf(TimeoutException.class)
+                        .max(Integer.MAX_VALUE)
+                        .delay(Delay.fixed(200, TimeUnit.MILLISECONDS))
+                        .doOnRetry(new Action4<Integer, Throwable, Long, TimeUnit>() {
+                            @Override
+                            public void call(Integer integer, Throwable throwable, Long aLong, TimeUnit timeUnit) {
+                                configProvider.stop().await();
```

Comment:
nevermind, I've decided to just add timeout and let it to bubble to the user as `BootstrapException`